In [59]:
import pandas as pd
import nltk
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
import string
from nltk.sentiment import SentimentIntensityAnalyzer

In [94]:
pd.set_option('display.max_colwidth', None)

In [60]:
nltk.download('vader_lexicon')

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\annegru\AppData\Roaming\nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


True

In [ ]:
comments = pd.read_excel(r"C:\Users\annegru\Liberty Global\Product Strategy and Planning - Data and insights\Products\Stream Relationship Survey\sentimentanalysis_comments.xlsx")
comments

In [62]:
comments.dropna(inplace=True)

In [ ]:
comments_list = comments['reason_behind_score'].tolist()
print(comments_list)

In [64]:
# tokenization
stoplist = set(stopwords.words('english'))
punctuation = string.punctuation + '‘’“”' 
ps = PorterStemmer()

def preprocessed_comments(comment_list):
    preprocessed_comments = []
    for comment in comment_list:
        comment = comment.lower()
        comment = ''.join([c for c in comment if c not in punctuation])
        tokens = word_tokenize(comment)
        tokens = [token for token in tokens if token not in stoplist and token not in punctuation and not token.isnumeric()]
        tokens = [ps.stem(token) for token in tokens]
        tokens = [token for token in tokens if token not in stoplist and token not in punctuation and not token.isnumeric()]
        preprocessed_comments.append(tokens)
    return preprocessed_comments

In [ ]:
print(preprocessed_comments(comments_list))

In [ ]:
preprocessed_comments_list = preprocessed_comments(comments_list)
print(preprocessed_comments_list)

In [ ]:
# initialize sentiment intensity analyzer
sid = SentimentIntensityAnalyzer()

# generate polarity scores (negative, neutral, positive, compound) per comment
for comment in comments_list:
    sentiment_scores = sid.polarity_scores(comment)
    print(f"Comment: {comment}")
    print(f"Sentiment Scores: {sentiment_scores}")
    print()

In [68]:
# Create empty lists to store the results
comment_list = []
compound_score_list = []

for comment in comments_list:
    sentiment_scores = sid.polarity_scores(comment)
    comment_list.append(comment)
    compound_score_list.append(sentiment_scores['compound'])

# Create a DataFrame from the lists
df = pd.DataFrame({'Comment': comment_list, 'Compound Score': compound_score_list})


In [ ]:
df

In [ ]:
zero_scores_comments = df.loc[df['Compound Score'] == 0.0, 'Comment']
zero_scores_comments

In [ ]:
positive_scores_comments = df.loc[df['Compound Score'] > 0.0]
positive_scores_comments

In [ ]:
negative_scores_comments = df.loc[df['Compound Score'] < 0.0]
negative_scores_comments

In [87]:
# bucketing comments into levels of negativity/positivity and neutral

print(f"Negative comments with compound score below 0: {negative_comments}")
print(f"Negative comments with compound score between 0-0.2: {neg_zero_to_two_comments}")
print(f"Negative comments with compound score between 0.2-0.4: {neg_two_to_four_comments}")
print(f"Negative comments with compound score between 0.4-0.6: {neg_four_to_six_comments}")
print(f"Negative comments with compound score between 0.6-0.8: {neg_six_to_eight_comments}")
print(f"Highly negative comments with compound score of 0.8 or below: {very_negative_comments}")


print(f"Neutral comments with compound score of 0: {neutral_comments}")


print(f"Positive comments with compound score above 0: {positive_comments}")
print(f"Positive comments with compound score between 0-0.2: {zero_to_two_comments}")
print(f"Positive comments with compound score between 0.2-0.4: {two_to_four_comments}")
print(f"Positive comments with compound score between 0.4-0.6: {four_to_six_comments}")
print(f"Positive comments with compound score between 0.6-0.8: {six_to_eight_comments}")
print(f"Highly positive comments with compound score of 0.8 or above: {very_positive_comments}")

Negative comments with compound score below 0: 1787
Negative comments with compound score between 0-0.2: 234
Negative comments with compound score between 0.2-0.4: 419
Negative comments with compound score between 0.4-0.6: 571
Negative comments with compound score between 0.6-0.8: 371
Highly negative comments with compound score of 0.8 or below: 192
Neutral comments with compound score of 0: 826
Positive comments with compound score above 0: 1808
Positive comments with compound score between 0-0.2: 183
Positive comments with compound score between 0.2-0.4: 332
Positive comments with compound score between 0.4-0.6: 585
Positive comments with compound score between 0.6-0.8: 470
Highly positive comments with compound score of 0.8 or above: 238


The 'compound' score is one of the sentiment scores provided by the VADER sentiment analyzer. It represents an overall sentiment polarity calculated using a specific algorithm in the VADER sentiment analysis tool. The compound score is computed based on heuristics used by VADER to determine the sentiment intensity of the text. It takes into account linguistic & grammatical rules as well as word order & context to calculate a normalized, weighted score. The score ranges from -1 (highly negative) to 1 (highly positive).

Out of 4420 comments in total, 1787 are considered negative as they have a compound score below 0. Out of those negative comments, 192 of them are considered highly negative as they have a compound score of -0.8 or less. 

Out of 4420 comments in total, 1808 are considered positive as they have a compound score above 0. Out of these positive comments, 238 are considered highly positive as they have a compound score of 0.8 or greater.

Out of 4421 comments in total, 826 are considered neutral because they have a compound score of 0.0